src/lock.rs

```diff
@@ -0,0 +1,103 @@
+use std::fs::write;
+use std::path::{Path, PathBuf};
+
+use serde::{Deserialize, Serialize, Serializer};
+
+use crate::{files, util::Fallible};
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+pub fn code_changed<P: AsRef<Path>>(lock_file: P) -> Fallible<bool> {
+    if lock_file.as_ref().is_file() {
+        let lock = Lock::read(lock_file)?;
+        let code_hashes = hash_files(lock.code_hashes.keys())?;
+        Ok(code_hashes != lock.code_hashes)
+    } else {
+        Ok(false)
+    }
+}
+
+pub fn write_lock<P: AsRef<Path>>(
+    lock_file: P,
+    source_files: &HashSet<PathBuf>,
+    code_files: &HashSet<PathBuf>,
+) -> Fallible {
+    let lock = Lock {
+        source_hashes: hash_files(source_files.iter())?,
+        code_hashes: hash_files(code_files.iter())?,
+    };
+    lock.write(&lock_file)
+}
+
+fn hash_files<'a>(files: impl Iterator<Item = &'a PathBuf>) -> Fallible<HashMap<PathBuf, String>> {
+    let mut code_hasher = Hasher::default();
+    files
+        .map(|p| match code_hasher.hash(p) {
+            Ok(hash) => Ok((p.clone(), hash)),
+            Err(err) => Err(err),
+        })
+        .collect::<Result<HashMap<_, _>, _>>()
+}
+
+/// Content for Yarner.lock files
+#[derive(Serialize, Deserialize)]
+struct Lock {
+    #[serde(serialize_with = "ordered_map")]
+    source_hashes: HashMap<PathBuf, String>,
+    #[serde(serialize_with = "ordered_map")]
+    code_hashes: HashMap<PathBuf, String>,
+}
+
+fn ordered_map<S, K, V>(value: &HashMap<K, V>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+    K: Ord + Serialize,
+    V: Serialize,
+{
+    let ordered: BTreeMap<_, _> = value.iter().collect();
```

Comment:
Why not put the `BTreeMap` direct inside `struct Lock`?

Reply:
But then, files are in order of insertion I think. We should then sort nevertheless...

Reply:
It seems I still need a serializer function to get "clean" paths.

Reply:
Couldn't you just do that in the place where you currently do
```
Ok(hash) => Ok((p.clone(), hash)),
```
?